File: example-servers/go/services/custom.go
package services

import (
	"net/http"
	"time"
	"fmt"
	"io"
)

func Chat(w http.ResponseWriter, r *http.Request) error {
	err, skipPreflight := ProcessIncomingRequest(&w, r)
	if err != nil { return err }
	if skipPreflight { return nil }

	// Text messages are stored inside request body using the Deep Chat JSON format:
	// https://deepchat.dev/docs/connect
	body, err := io.ReadAll(r.Body)
	if err != nil { return err }

	fmt.Println("Request Body:", string(body))


	// Create response
	jsonResponse, err := CreateTextResponse(w, "This is a response from the Go server. Thank you for your message!")
	if err != nil { return err }

	// Send response
	w.Header().Set("Content-Type", "application/json")
	w.Write(jsonResponse)
	return nil
}

func ChatStream(w http.ResponseWriter, r *http.Request) error {
	err, skipPreflight := ProcessIncomingRequest(&w, r)
	if err != nil { return err }
	if skipPreflight { return nil }
	
	// Text messages are stored inside request body using the Deep Chat JSON format:
	// https://deepchat.dev/docs/connect
	body, err := io.ReadAll(r.Body)
	if err != nil { return err }

	fmt.Println("Request Body:", string(body))
	
	respondWithStream(w)
	return nil
}

func respondWithStream(w http.ResponseWriter) {
	w.Header().Set("Content-Type", "text/event-stream")
	flusher := w.(http.Flusher)
	responseChunks := []string{"This", "is", "a", "response", "from", "the", "Go", "server", ".", "Thankyou", "for", "your", "message!"}
	ticker := time.NewTicker(70 * time.Millisecond)
	chunkIndex := 0

	outerloop:
		for {
			select {
				case <-ticker.C:
					// create response
					jsonResponse, err := CreateTextResponse(w, responseChunks[chunkIndex] + " ")
					if (err != nil) { break outerloop }
					// Send response
					w.Write([]byte("data: " + string(jsonResponse) + "\n\n"))
					chunkIndex += 1
					flusher.Flush()
				default:
					if chunkIndex == len(responseChunks) {
						break outerloop
					}
			}
		}

	ticker.Stop()
}

func Files(w http.ResponseWriter, r *http.Request) error {
	err, skipPreflight := ProcessIncomingRequest(&w, r)
	if err != nil { return err }
	if skipPreflight { return nil }

	// Files are stored inside a form using Deep Chat request FormData format:
	// https://deepchat.dev/docs/connect
	err = r.ParseMultipartForm(32 << 20) // Max memory of 32MB
	if err == nil {
		files, ok := r.MultipartForm.File["files"]
		if ok {
			for _, file := range files {
				fileName := file.Filename
				fmt.Println("File name:", fileName)
			}
		}
		// If there are text messages sent along with the files, they will also be in the form
		for key, values := range r.Form {
			// Skip the "files" property
			if key == "files" { continue }

			for _, value := range values {
				fmt.Println("Message", value)
			}
		}
	} else {
		// If message is sent without any files, the text will be stored inside body using Deep Chat JSON request format:
		// https://deepchat.dev/docs/connect
		body, err := io.ReadAll(r.Body)
		if err != nil { return err }

		fmt.Println("Request Body:", string(body))
	}

	// Create response
	jsonResponse, err := CreateTextResponse(w, "This is a response from the Go server. Thank you for your message!")
	if err != nil { return err }

	// Send response
	w.Header().Set("Content-Type", "application/json")
	w.Write(jsonResponse)
	return nil
}

File: example-servers/node/express/src/services/custom.ts
import {Request, Response} from 'express';

export class Custom {
  public static async chat(body: Request['body'], res: Response) {
    // Text messages are stored inside request body using the Deep Chat JSON format:
    // https://deepchat.dev/docs/connect
    console.log(body);
    // Sends response back to Deep Chat using the Response format:
    // https://deepchat.dev/docs/connect/#Response
    res.json({text: 'This is a respone from ExpressJs server. Thankyou for your message!'});
  }

  public static async chatStream(body: Request['body'], res: Response) {
    // Text messages are stored inside request body using the Deep Chat JSON format:
    // https://deepchat.dev/docs/connect
    console.log(body);
    const responseChunks = 'This is a respone from ExpressJs server. Thankyou for your message!'.split(' ');
    res.setHeader('Content-Type', 'text/event-stream');
    res.setHeader('Cache-Control', 'no-cache');
    res.setHeader('Connection', 'keep-alive');
    res.setHeader('Access-Control-Allow-Origin', '*');
    Custom.sendStream(res, responseChunks);
  }

  private static sendStream(res: Response, responseChunks: string[], chunkIndex = 0) {
    setTimeout(() => {
      const chunk = responseChunks[chunkIndex];
      if (chunk) {
        // Sends response back to Deep Chat using the Response format:
        // https://deepchat.dev/docs/connect/#Response
        res.write(`data: ${JSON.stringify({text: `${chunk} `})}\n\n`);
        Custom.sendStream(res, responseChunks, chunkIndex + 1);
      } else {
        res.end();
      }
    }, 70);
  }

  public static async files(req: Request, res: Response) {
    // Files are stored inside a form using Deep Chat request FormData format:
    // https://deepchat.dev/docs/connect
    if (req.files as Express.Multer.File[]) {
      console.log('Files:');
      console.log(req.files);
    }
    // When sending text along with files, it is stored inside the request body using the Deep Chat JSON format:
    // https://deepchat.dev/docs/connect
    if (Object.keys(req.body).length > 0) {
      console.log('Text messages:');
      // message objects are stored as strings and they will need to be parsed (JSON.parse) before processing
      console.log(req.body);
    }
    // Sends response back to Deep Chat using the Response format:
    // https://deepchat.dev/docs/connect/#Response
    res.json({text: 'This is a respone from ExpressJs server. Thankyou for your message!'});
  }
}
